### DISCUSSION TITLE
Frame buffer and live recording update based on fixed frame rate

### DISCUSSION THREAD

**Discussion by jacopoabramo** [ImSwitch/ImSwitch#91] [Ideas]
I've been working on integrating a Ximea camera which is currently used in our setup, and to better understand the DetectorManager structure I've looked into the code of the Manager itself. I especially focused on the method `updateLatestFrame`, currently implemented as follows:

```python
    def updateLatestFrame(self, init):
        """ :meta private: """
        try:
            self.__image = self.getLatestFrame()
        except Exception:
            self.__logger.error(traceback.format_exc())
        else:
            self.sigImageUpdated.emit(self.__image, init)
```

Now I may be misunderstanding what happens here, but from what I'm guessing is that every new acquired frame is directly sent to the live view through the signal `sigImageUpdated`. This is a system I've also implemented in the past for other projects, but I realized after a bit of digging that is not particularly efficient for some reasons:

- updating the live viewer everytime a new image is available may cause ImSwitch to freeze in case new frames are available at a faster rate than what ImSwitch itself can actually support (I experienced this with my live recording plugin for napari using the Ximea camera)
- updating the live viewer at everyframe is not really necessary in itself, as most monitors can support 60 Hz refresh rate, or 30 Hz for cheaper ones (... or 144 Hz for gaming monitors but I don't think it's something you can find in a microscope setup anyway).
- the signal/slot system implemented in Qt is very useful but not very efficient to perform direct transmission of data in a stream-like way. Memory buffers implemented as FIFOs are more reliable to make sure no frames are lost.

For this reason I would propose the following integration:

- a frame buffer for each detector manager of the setup in order to keep track of the acquired frames;
- integrating a QTimer which would update the live view every 30/60 Hz instead of directly updating the live view with a frame stored in the frame buffer.

Using a frame buffer can be particularly useful in case you want to do live recordings and real-time post processing on the frames an user has captured. I don't know if this is the way it's already been done for non-live recordings, and again I may be missing something in the way ImSwitch handles this.

**Comment by xcasas** [comment]
Hi @jacopoabramo , there is actually a thread that runs every `updatePeriod` and calls the `updateLatestFrame`. The thread is in [DetectorsManager, line 155](https://github.com/kasasxav/ImSwitch/blob/98d9232d76608c9617f2ad540a0e600d64a3b63e/imswitch/imcontrol/model/managers/DetectorsManager.py) and it already handles all the detectors at once depending on which ones are active for display. 
The updatePeriod is hardcoded to 100ms in [MasterController, line 30](https://github.com/kasasxav/ImSwitch/blob/b1d51943203cbcf357ecdba85cae5ab71bfd260a/imswitch/imcontrol/controller/MasterController.py), but it should be possible to add as a parameter and maybe change from the config file for example.
There is no buffer though, mostly because our camera interface already had one and the function `getLatestFrame` just gets the last image in the buffer at the time it's executed. But if you think there is a better way let me know.
